src/Symfony/Component/Security/Guard/AuthenticatorInterface.php

```diff
@@ -91,7 +91,7 @@ public function getUser($credentials, UserProviderInterface $userProvider);
      *
      * @param mixed $credentials
      *
-     * @return bool
+     * @return mixed true if the credentials are valid, anything else otherwise.
```

Comment:
This one really bothers me. Having a return type declaration `bool` on this method could really be a good guidance for implementers. And I don't really see the benefit in allowing implementations to anything (as the documentation suggests) to make the authentication fail. I'd suggest to deprecate non-boolean return values in 4.4, so we can add a `bool` return type declaration eventually.

Reply:
/cc @weaverryan

Reply:
@derrabus maybe we should open a PR to fix the cases where we test that a non-boolean works, to trigger feedback at least?

Reply:
The corresponding code is here: https://github.com/symfony/symfony/blob/5255ab89d2f08e8a929ee6e2a02e605cc3ac588b/src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php#L116-L118

Test case: https://github.com/symfony/symfony/blob/bee347c6f2a8324e94a40953d98ff056f567b246/src/Symfony/Component/Security/Guard/Tests/Provider/GuardAuthenticationProviderTest.php#L90-L119

We could patch the test case to have the mock return `false` instead of `null`.

Reply:
> Having a return type declaration bool on this method could really be a good guidance for implementers. And I don't really see the benefit in allowing implementations to anything (as the documentation suggests) to make the authentication fail.

I agree with that. The whole guard interface was made to be super user-friendly... and in some cases, maybe unnecessarily so. We wanted to force the user to `return true` to explicitly say "yes, the credentials are good". But if they forget to return anything, that's "falsy" anyways.